Fix color column position when `linebreak` is set

matrdr opened the pull request:
As reported in #16837 the 'colorcolumn' option does not behave as
expected, when the 'linebreak' option is set. Instead of always
highlighting the same character independent of the window's width, the
highlighted character changes with changing window sizes.

Furthermore, the color column ignored the inserted "virtual text" that
comes with the 'showbreak' option. This resulted in the color column
possibly highlighting the text inserted by the 'showbreak' option
instead of actual, user inserted text. Now, this "virtual text" is
ignored when determining the color column's position.

matrdr committed:
Fix `colorcolumn` when `linebreak` is on

The color column used to change its position when `linebreak` is on and
the window is resized. The expected behaviour for a color column would
be that it is always on the same character no matter how often the line
wraps.

An own variable was introduced that tracks the amount of characters that
were inserted due to the `linebreak` and/or the `breakindent` option.
This is used to evaluate the position of the color column.

There is still an issue that the end of the screen line gets highlighted
for some reason.

matrdr committed:
Fix color column for leading tabs

matrdr committed:
Fix `linebreak` + `colorcolumn` edge case

When the color column was exactly on the character that is the last
character before the break, the color column would be set incorrectly.

This only leaves tests regarding `cursorcolumn` and `showbreak` failing.

matrdr committed:
Fix showbreak and cursorcolumn

This changes the behaviour of how the showbreak option is treated in
combination with color column. Previously the color column could also be
displayed on the "virtual text" of the showbreak text.

As this is not what the color column is supposed to be, the color column
now skips the showbreak columns and only displays in actual text or
after the actual text.

matrdr committed:
Change expected 'conceal' test results

Two tests that use the `conceal`, `cursorcolumn`, `linebreak`,
`showbreak` and `colorcolumn` feature failed, due to changes in how the
color column is placed.

Previously, the color column used to change its position based on the
window width and did not respect the `showbreak` characters. As the
highlighted character is now always consistent, regardless of the
window's width, the test now reflects this changed behaviour.

Signed-off-by: Matthias Rader <matthias.rader@outlook.com>

habamax commented:
Could you show a screenshot with several long lines (~200 chars) and `set colorcolumn=20,40,60,80,100,120`?

![image](https://github.com/user-attachments/assets/1921025e-9a77-447d-a57c-553855c04318)

habamax commented:
oh, nevermind I see what you try to fix now!

chrisbra commented:
converting to draft for now. Thanks

matrdr committed:
Change expectation in color column test

matrdr committed:
Fix showbreak text being "colorcolumned"

When the first character of a line was to be marked with a color column
and the showbreak option was set to some text, the last character of the
showbreak line was highlighted in addition to the first "real" character
in the line.

By specifically stating that the currently drawn characters are "extra
character", this prevents them from being "colorcolumned".

matrdr committed:
Fix indentation

matrdr committed:
Change expected result in colorcolumn test case

The previously expected result had parts of the showbreak characters
marked with the colorcolumn highlight. This is not really an expected
behaviour for a color column.

As the color column now does not highlight these non-text elements any
more, the expected test result has been altered.

chrisbra commented:
this has stalled, so closing
